solr/core/src/java/org/apache/solr/util/ModuleUtils.java

```diff
@@ -0,0 +1,128 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.solr.util;
+
+import org.apache.solr.common.StringUtils;
+import org.apache.solr.common.util.StrUtils;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.lang.invoke.MethodHandles;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.Set;
+import java.util.regex.Pattern;
+import java.util.stream.Collectors;
+
+/**
+ * Parses the list of modules the user has requested in solr.xml, property solr.modules or environment SOLR_MODULES.
+ * Then resolves the lib folder for each, so they can be added to class path.
+ */
+public class ModuleUtils {
+  private static final Logger log = LoggerFactory.getLogger(MethodHandles.lookup().lookupClass());
+  public static final String MODULES_FOLDER_NAME = "contrib"; // TODO Change to "modules"
+  private static final Pattern validModNamesPattern = Pattern.compile("[\\w\\d-_]+");
+
+  /**
+   * Returns a path to a module's lib folder
+   * @param moduleName name of module
+   * @return the path to the module's lib folder
+   */
+  public static Path getModuleLibPath(Path solrInstallDirPath, String moduleName) {
+    return getModulesPath(solrInstallDirPath).resolve(moduleName).resolve("lib");
+  }
+
+  /**
+   * Finds list of module names requested by system property or environment variable
+   * @return set of raw volume names from sysprop and/or env.var
+   */
+  static Set<String> resolveFromSyspropOrEnv() {
+    // Fall back to sysprop and env.var if nothing configured through solr.xml
+    Set<String> mods = new HashSet<>();
+    String modulesFromProps = System.getProperty("solr.modules");
+    if (!StringUtils.isEmpty(modulesFromProps)) {
+      mods.addAll(StrUtils.splitSmart(modulesFromProps, ',', true));
+    }
+    String modulesFromEnv = System.getenv("SOLR_MODULES");
```

Comment:
Please note - I made a "shortcut" here. Traditionally we have parsed environment variables in `bin/solr` and converted them to system properties on command line `java -jar start.jar -Dfoo=bar`, and then pick up *only* the sysprop in some xml config file.

In this patch I wanted to support both system property and env.var for all 9.0 users, even if they bring in a custom `solr.xml`, (without the `modules` tag). Thus we first look for a value from `solr.xml` (which also picks up direct sysprop such as `bin/solr start -Dsolr.modules=foo` or `bin/solr start -a "-Dsolr.modules=foo"`. But we also fall back to parsing sysprop and env.var direclty in Java, if no or empty value from solr.xml. But if a user in  a custom `solr.xml` choose to override or pull the value from another sysprop, that will always win. Direct sysprod+env is a fallback.

Reply:
Yeah, being able to add stuff with env vars in preference to editing config files makes sense in our containerized world.

Reply:
Glad to see some support. In the future I want to slim down `bin/solr`, `bin/solr.cmd` and also `SolrCLI` to remove all kinds of boiler-plate code to map between env.var and sys.prop, and introduce a new `ConfigResolver` to Solr, so some piece of code could say `ConfigResolver.getInt("solr.number")` and it would return and convert the value from either sys.prop `solr.number` or env.var `SOLR_NUMBER`. We could also add stuff like `ConfigResolver.getListOfStrings("solr.modules")` and it would return a `List<String>` for you. It could even support comma-separated, semicolon-separated or why not also props of format `["foo","bar"]`? 

If we can somehow tie it in with loading values from solr.xml and our other config files, then we could have a DSL that would be really pleasant to use for falling back from explicit xml-config value to any of sys.prop or env.var.

Of course we still need to parse and set JVM-level props like memory in bin/solr.

Reply:
Ok, this is off topic here, so I made https://issues.apache.org/jira/browse/SOLR-15960 to track.